### DISCUSSION TITLE
Bundling 3rd Party Dependencies

### DISCUSSION THREAD

**Discussion by rma-tom** [CVWRSM/cvwrsm#104] [General]
This is the discussion that @dezidizon said would be started in issue #103 
We'll require @hxiedwr and @jmdegeorge to participate, and I think that @adamkorynta should at least be aware of the discussion.

In the current Eclipse build environment and in the distribution packages of WRIMS there are quite a few third-party jars and DLLs and many of them are duplicated, which can be problematic. We want to clean that up for devOps.

Josh is setting up the Gradle builds for the plugin jars for WRIMS, so we need to provide him with explicit directions for what belongs in the jars in the dropins folder, which are currently these:
- gov.ca.dwr.hecdssvue_<version>.jar
- gov.ca.dwr.jdiagram_<version>.jar
- gov.ca.dwr.wresl.xtext.editor.ui_<version>.jar
- gov.ca.dwr.wresl.xtext.editor_<version>.jar
- wrimsv2_plugin_<version>.jar

Third-party dependencies include solvers, HEC DSS and HDF5 libraries, and utilities like JDiagram, flogger and so on. In a recent discussion at RMA we identified three ways these can be located in the WRIMS directory structure: 
- inside plugin jars
- in folders in the file system
- in jars-of-jars like the HEC dependencies jar, which may or may not be a plug

## Easy stuff
- The wresl editor jars don't contain any third-party contents, so they're trivial.

- The JDiagram jar contains one third-party jar, JDiagram.jar. Dependency analysis with jdeps and tattletale hasn't identified any dependencies on any of the jdiagram classes (members of packages named com.mindfusion.*) outside of the gov.ca.dwr.jdiagram package. There's a copy of JDiagram.jar in the lib folder in the WRIMS project directory, but I didn't see any effect on the WRIMS GUI when I deleted it. My recommendation is that we keep the JDiagram jar inside the gov.ca.dwr.jdiagram plugin and remove it from other locations. Because the mindfusion.com products are commercial and proprietary, they may not be available from public repositories, so we'll have to figure out how gradle will find that jar.

## Harder stuff 1 -- the wrims_plugin jar
The wrimsv2_plugin jar includes folders named lib and lib_x64 that contain a number of jars and DLLs. So far, our dependency analysis has found that the contents of these folders are only used by the WRIMS compute engine -- contained in the the WRIMSv2 or wrims-core jars. Since the compute engine is actually run from a batch file that the GUI builds, and that batch file depends on copies of those jars and DLLs that are in the lib folder, the copies of those DLLs and jars in the wrims_plugin jar are never used, and we can greatly simplify the building and maintenance of the wrims_plugin jar by leaving the lib and lib_x64 folders out of the wrims_plugin jar.

This would be a significant change to the jar contents, and DWR has to approve the change. If we don't make these changes, though, we have to develop a gradle process for assembling those contents for inclusion in the jars.

## Harder stuff 2 -- the hecdssvue jar
The DWR hecdssvue jar also contains a lot of jars and DLLs that don't appear to be used. I haven't completed tracing out the dependencies inside the jar, but it looks like there are real dependencies on some calsim GUI components, some RMA classes, and a vista package.

As with the wrimsv2_plugin, I favor reducing the amount of unused contents in this plugin jar, so that we don't unnecessarily complicate the gradle build and ongoing maintenance of the package. Again, it's a big change in the contents of the jar, so DWR has to be aware and approve of them.

**Comment by hxiedwr** [comment]
Hi Tom,

I agree on most of the ideas you have.

For the WRIMSv2_Plugin jar, some of the jars like WRIMSv2.jar and related jars are needed inside the plugin. There are some functions in WRIMS 2 GUI using WRIMSv2.jar. I can show you sometime later.

Thank you,

Hao
________________________________
From: Tom Evans ***@***.***>
Sent: Thursday, July 11, 2024 1:56:20 PM
To: CentralValleyModeling/wrims ***@***.***>
Cc: Xie, ***@***.*** ***@***.***>; Mention ***@***.***>
Subject: [CentralValleyModeling/wrims] Bundling 3rd Party Dependencies (Discussion #104)


This is the discussion that @dezidizon<https://github.com/dezidizon> said would be started in issue #103<https://github.com/CentralValleyModeling/wrims/issues/103>
We'll require @hxiedwr<https://github.com/hxiedwr> and @jmdegeorge<https://github.com/jmdegeorge> to participate, and I think that @adamkorynta<https://github.com/adamkorynta> should at least be aware of the discussion.

In the current Eclipse build environment and in the distribution packages of WRIMS there are quite a few third-party jars and DLLs and many of them are duplicated, which can be problematic. We want to clean that up for devOps.

Josh is setting up the Gradle builds for the plugin jars for WRIMS, so we need to provide him with explicit directions for what belongs in the jars in the dropins folder, which are currently these:

  *   gov.ca.dwr.hecdssvue_.jar
  *   gov.ca.dwr.jdiagram_.jar
  *   gov.ca.dwr.wresl.xtext.editor.ui_.jar
  *   gov.ca.dwr.wresl.xtext.editor_.jar
  *   wrimsv2_plugin_.jar

Third-party dependencies include solvers, HEC DSS and HDF5 libraries, and utilities like JDiagram, flogger and so on. In a recent discussion at RMA we identified three ways these can be located in the WRIMS directory structure:

  *   inside plugin jars
  *   in folders in the file system
  *   in jars-of-jars like the HEC dependencies jar, which may or may not be a plug

Easy stuff

  *   The wresl editor jars don't contain any third-party contents, so they're trivial.

  *   The JDiagram jar contains one third-party jar, JDiagram.jar. Dependency analysis with jdeps and tattletale hasn't identified any dependencies on any of the jdiagram classes (members of packages named com.mindfusion.*) outside of the gov.ca.dwr.jdiagram package. There's a copy of JDiagram.jar in the lib folder in the WRIMS project directory, but I didn't see any effect on the WRIMS GUI when I deleted it. My recommendation is that we keep the JDiagram jar inside the gov.ca.dwr.jdiagram plugin and remove it from other locations. Because the mindfusion.com products are commercial and proprietary, they may not be available from public repositories, so we'll have to figure out how gradle will find that jar.

Harder stuff 1 -- the wrims_plugin jar

The wrimsv2_plugin jar includes folders named lib and lib_x64 that contain a number of jars and DLLs. So far, our dependency analysis has found that the contents of these folders are only used by the WRIMS compute engine -- contained in the the WRIMSv2 or wrims-core jars. Since the compute engine is actually run from a batch file that the GUI builds, and that batch file depends on copies of those jars and DLLs that are in the lib folder, the copies of those DLLs and jars in the wrims_plugin jar are never used, and we can greatly simplify the building and maintenance of the wrims_plugin jar by leaving the lib and lib_x64 folders out of the wrims_plugin jar.

This would be a significant change to the jar contents, and DWR has to approve the change. If we don't make these changes, though, we have to develop a gradle process for assembling those contents for inclusion in the jars.

Harder stuff 2 -- the hecdssvue jar

The DWR hecdssvue jar also contains a lot of jars and DLLs that don't appear to be used. I haven't completed tracing out the dependencies inside the jar, but it looks like there are real dependencies on some calsim GUI components, some RMA classes, and a vista package.

As with the wrimsv2_plugin, I favor reducing the amount of unused contents in this plugin jar, so that we don't unnecessarily complicate the gradle build and ongoing maintenance of the package. Again, it's a big change in the contents of the jar, so DWR has to be aware and approve of them.

**Comment by rma-tom** [comment]
Hi Hao (@hxiedwr)

Here's where I've gotten to on dependencies for the WRIMSv2_plugin jar as we will build it in gradle.

Where WRIMSv2_plugin relies on classes in WRIMSv2.jar (now wrims-core.jar) we have two ways of providing those classes. 
1. Put wrims-core.jar inside wrimsv2_plugin.jar and include it in the plugin's Bundle-ClassPath
2. Put wrims-core.jar on the classpath for the WRIMS GUI application and include the needed packages in the plugin's Import-Package list

The Eclipse builds have used the first approach. For reliability, I suggest we do the same in the gradle builds.

These are the direct dependencies that I've identified to be included in the wrimsv2_plugin jar:
- wrims-core
- dsm2-input-model
- vista
- jfreechart
- apache commons

We also need the HEC packages, but they have their own plugin, so we shouldn't have to include them in wrimsv2_plugin.

The wrims-core packages have additional dependencies, which are identified in its gradle build. I suggest that we duplicate most of the wrims-core dependencies within the wrims2_plugin bundle. It doesn't cost much besides a modest increase in file sizes and saves time and worry. We can guarantee that the versions in the plugin and the WRIMS engine runs will be the same because they'll be pulled from the same sources at build time. I do think we can omit the LP solvers (coinor.jar, com.google.ortools.linearsolver.jar, lpsolve55j.jar, XAOptimizer.jar, jCbc.dll, lpsolve55.dll, lpsolve55j.dll, xav16_x64.dll) from the wrims_plugin jar because WRIMS doesn't actually run computes from the plugin. I assume that we can also omit 32-bit versions of jars and DLLs from our gradle builds.

Hao, if you approve of this set of contents for the plugin, I can work with Josh (@jmdegeorge) on identifying specific versions and repositories, and he can create the gradle/buildship setups to automate the build.

Notes on the DWR hecdssvue jar will follow soon.

**Comment by hxiedwr** [comment]
Hi Tom,

I agree your way of handling lpsolve and 32 bit jars.

On the other hand, I still think wrims-core.jar has to be in both “lib” folder and wrimsv2_plugin.jar file.

One more thing I would suggest is to follow the name convention and use gov.ca.dwr.wrims or gov.ca.dwr.wrimsgui as the folder and plugin name. The way of using wrimsv2_plugin as folder and plugin name is not consistent with plugin developments widely.

Thank you,

Hao
________________________________

**Comment by rma-tom** [comment]
Hi Hao

I agree with you about putting wrims-core.jar in both places. Sorry if that wasn't clear. I don't think we're worried about the additional file size, and since we're building the jars and the program installation, we can guarantee that the version of wrims-core is the same in both places.

I also agree regarding the names of the jars. That should be straightforward to implement.

Cheers
Tom

**Comment by hxiedwr** [comment]
Thank you,

Hao
________________________________

**Comment by rma-tom** [comment]
Hi Everyone (@hxiedwr, @jmdegeorge, @adamkorynta)

Regarding the contents of the gov.ca.dwr.hecdssvue_<version>.jar:

- As with the wrimsgui jar, we can omit the HEC dependencies from this plugin jar because they'll be provided by their own plugin. This modification has already been done in the HEC-DSS7 branch of the code, we need to assure that the same code is in the devOps branch.
- We will keep calsimgui.jar, calsim.jar, and vista.jar as they are provided by DWR. 
- As with wrimsgui, there's no need to duplicate 32-bit and 64-bit dependencies
- Moving the third-party dependencies in this bundle to recent versions that can be pulled from repositories will require some updates. The modules in the swixml and jdom jars are available in newer versions in Maven. The jpython package should be updated to jython, which was renamed in 1999.
- The contents of the COM jar are very old. It does have third-party contents, but I doubt that any of them can be found in current repositories. We will probably want to keep this in our own artifact space until its contents can be retired.

**Comment by rma-tom** [comment]
Moving from an email discussion to here, so it doesn't get lost.

On 9 August, I provided a test version of the plugin wrimsv2_plugin.jar to @hxiedwr for testing. 

Hao found the following problems:
1. The menu item of Data->Load Dss/Study doesn’t work.
This menu item is used when you debug the model by running the model as “debug as” and pause the model in the middle of the run. It is basically to load the selected dss files to make comparisons in the Variables and Watch View on the right hand side. I guess some HecLib 7 is not included in the wrimsv2 plugin jar file.
2. The popup menu items for a main file (Export Study, Cycle WRESL, and Study WRESL) or a launch file (Export Study) fail. I think the reason may be related to wrimsv2.jar which is used to parse the main file and study. Since the name of the wrimsv2.jar changes to wrims-core…., there is a confusion for the plugin.

On 14 August, I did a few more tests, particularly following Hao's suggestion that the name of the WRIMS jar that's inside the WRIMS plugin jar might have an effect.

I revised the wrimsv2_plugin jar, renamed the wrims-core.jar that was inside the plugin to “WRIMSv2.jar” – so it’s the new jar, but with the old name – and that appeared to resolve the issue with the popup menu. We’ll need to figure out where the jar name is explicitly referenced – it’s puzzling that the name of the jar only seems to matter for this one action.
 
I also found that the Data->Load Dss/Study function works (on my computer, anyway) but only if the loaded DSS files are version 6. I think this indicates that the problem is not with the contents of the wrimsv2_plugin jar itself – since it no longer contains DSS libraries – but either with the DSS libraries in the hec dependencies jar, or possibly with the case of the path parts: “c_kswck” vs. “C_KSWK” for example.

I shared this with Hao, and we exchanged the following messages.

Hao:
For the first issue, I guess you can change the wrimsv2.jar to wrims-core at the “Bundle-ClassPath” in the “Meta-Inf\Manifest.mf” file.
 
For the second issue, I think we have to include HecLib 7 libraries in the plugin jar file to replace HecLib 6.

Tom:
As I’ve been building these test jars, I’ve kept the jar names in the “Bundle-ClassPath” of manifest file in sync with the names of the jars in the lib directory inside the jar. The WRIMs jar inside the wrims_plugin jar was “wrims-core-Feature-wrims-devops-SNAPSHOT.jar” in the lib_x64 folder in the 9 Aug jar, and “WRIMv2.jar” in both places in the jar I revised yesterday. When the name in both places is “WRIMSv2.jar” the popup menu actions work, but when it’s “wrims-core…” the popup actions don’t work. The jar was only renamed; its contents were not altered. I’m surprised that the jar name has this effect and I’m going to talk with Adam and Josh (@adamkorynta @jmdegeorge ) to get their ideas on why this should be the case.

On your second point, if we are moving forward to HEClib 7 libraries, should be be using the hec-dependencies jar? The GUI application I was using as a test for the revised plugin jar has moved the DSS dependencies to that separate jar (hec-dependencies_1.0.0.0.jar, which is also a plugin). The copy of that jar that I have is from last year, which may be part of the problem. Is there a newer build of that jar somewhere in GitHub, or somewhere else accessible?

Hao:
Yes, it would be great to use the hec_dependencies jar if possible. Otherwise, we have to include the heclib 7 in the plugin.
I think you can build the latest heclib 7 plugin from your computer set up. A problem with that is you may have to change the code to work with the latest library.

**Comment by hxiedwr** [comment]
This issue reminds me a possible solution. What you probably need to do is: launch WRIMS 2 GUI at the development environment and close it down. Then you copy the folders of "configuration", "features", "p2", and "plugins" from the Eclipse Luna, which provides the IDE for WRIMS development, to the WRIMS 2 GUI package where you have dropins updated.

**Reply by rma-tom** [reply]
Hi Hao -- Thank you for the suggestion and the encouragement. I tried your idea and didn't succeed in getting the problems resolved. However, based on the results we saw with the revisions I described, I think that we've demonstrated that the jars included in the package are very likely complete. Rather that spend more time trying to get the example to run, I think we should have Josh proceed with building a jar with those contents in a more consistent environment.

**Comment by hxiedwr** [comment]
I still think this is a potential problem. The behavior of Eclipse is not consistent with dropins. Most of time, I just need place the new dropins to the "dropins" folder. Other times, I have to copy additional folders to the GUI package from Eclipse. That also makes me hesitate with creating a method to auto build WRIMS.

In your case, the last resort I would take is to build the WRIMS GUI based on the Eclipse package. Copy dropins and additional folders not available in the Eclipse from the WRIMS GUI to Eclipse, and modify some meta data files in Eclipse. Then make Eclipse the new WRIMS GUI.

Thank you,

Hao

**Comment by hxiedwr** [comment]
Hi Tom,

I am wondering if you tried to use “wrimsv2_gui_start.bat” file to start WRIMS 2 GUI instead of “WRIMSv2_GUI_x64.exe” file after all the changes you made.

This approach may help Eclipse recognize the jar file with the changed name inside the plugin.
________________________________
From: Xie, ***@***.***
Sent: Tuesday, August 20, 2024 3:49:49 PM
To: CentralValleyModeling/wrims ***@***.***>; CentralValleyModeling/wrims ***@***.***>
Cc: Mention ***@***.***>
Subject: RE: [CentralValleyModeling/wrims] Bundling 3rd Party Dependencies (Discussion #104)


I still think this is a potential problem. The behavior of Eclipse is not consistent with dropins. Most of time, I just need place the new dropins to the “dropins” folder. Other times, I have to copy additional folders to the GUI package from Eclipse. That also makes me hesitate with creating a method to auto build WRIMS.



In your case, the last resort I would take is to build the WRIMS GUI based on the Eclipse package. Copy dropins, copy additional folders not available in the Eclipse from the WRIMS GUI, and modify some meta data files.



Thank you,



Hao

**Reply by rma-tom** [reply]
Hi Hao

Excellent suggestion! Running from the batch file re-initializes things differently from running from the EXE. It also puts a great deal of output into a terminal window, where I can see it.

We needed an antlr runtime jar in the wrimsv2_plugin jar to support wrims-core. I'm not sure why renaming wrms-core to WRIMSv2 worked around that problem, but it's not required once the antlr dependency is resolved.

I've got a much better jar now, which I will make available to you.

**Comment by hxiedwr** [comment]
The problem of loading dss for comparison function not working might be related to mix reading HecLib 6 and 7 dss. In the test case we have, dv file to load for comparison is HecLib 7 dss but sv file is HecLib 6 dss.

When loading them together, I don’t know how HecLib 7 will distinguish the difference. I hope to have the code in WRIMS to detect which version the dss file is and then read the dss accordingly.

________________________________

**Comment by rma-tom** [comment]
I had wondered if the loader was trying to find
/CALSIM/C_KSWCK/FLOW-CHANNEL/01Jan1920/1MON/L2020A/
and failing to recognize
/CALSIM/c_kswck/flow-channel/01Jan1920/1Month/L2020A/
as being the same thing.

Have we (or HEC) developed a bit of code that will find both of those appropriately in 6 or 7 files, if asked to search for either of those paths? That seems like a generally useful thing.

**Comment by hxiedwr** [comment]
What I found is HecLib7 can load both v6 or both v7 dss files with our current code but gets confused when v6 and v7 loaded together. If v6 and v7 are read in this sequence, only v6 gets read but v7 fails, vice verse.

The same goes to the report tool. We have to have the code to make HecLib 7 recognize which version is the dss file first and then read the file accordingly.
________________________________

**Reply by rma-tom** [reply]
Should we make a ticket for this as an issue? It's a little off-topic in this discussion

**Comment by hxiedwr** [comment]
I agree.
